Fix color validation by avoiding null vectors between identical colors & running all checks concurrently

robpvn opened the pull request:
This is a fix for issue #16837 , which also has some (possibly) unintended consequences worth discussing.

If the picked color and a forbidden color are identical (like if they both picked the same palette color and in the special case when a picked color is outside of the allowed range and the method
returns the picked color as the forbidden color), the vector between them is zero and the maths for adjusting the color fails by hitting the iteration limit. This changes the zero vector to the smallest possible vector in order to avoid the issue.

I also added an extra message if you hit the iteration failsafe that picks a random colour. 

However, this can result in some seriously close adjustments in the case of picking identical palette colors, which might be undesirable compared to picking a new palette color. (See picture of adjusted identical palette picks below.) While technically the current behaviour of picking a random new palette color is a bug, I have a feeling many people might see it as a feature. Comments?

![](http://i.imgur.com/M01bRhe.png)

robpvn committed:
Added error message if color could not be adjusted

The failsafe in ColorValidator aborts after 255 iterations of
adjusting the color and picks a random color. This message makes it
clearer to the user. Results in two messages being displayed,
first one about adjusting and the about a random color pick.

robpvn committed:
Avoid null vectors when making colors valid

If the picked color and a forbidden color are identical (like
if they both picked the same palette color and in the special case
when a picked color is outside of the allowed range and the method
returns the picked color as the forbidden color),
the vector between them is zero and the maths for adjusting
the color fails by hitting the iteration limit. This changes
the zero vector to the smallest possible vector in order to
avoid the issue.

This can result in some seriously close adjustments in the case of
picking identical palette colors, which might
be undesirable compared to picking a new palette color.

robpvn commented:
Woops, this actually only fixes the first part of issue #16837, where the range adjustment didn't get done. The second part, where the two checks aren't run concurrently, isn't fixed yet. I can look at that after I get feedback on the identical palette pick question.

robpvn commented:
Now I've made a fix for the whole issue #16837, which ensures that all checks are made at the same time.

abcdefg30 commented:
Sorry for looking at this only so late. The code changes look good to me on a first glance.

I'd agree that picking a random new colour is worse than just adjusting the colour slightly, so the behaviour in this PR should be fine.

One other thing I noticed: You should remove the `allForbidden` variable now that it is unused.

robpvn commented:
Don't worry about responding late, it turns out I'm slow to respond as well! :D 

I deleted the unused variable and rebased the commits to the newest bleed.

abcdefg30 reviewed (dismissed):
Code changes look good to me and I didn't spot any regressions.

abcdefg30 commented:
I opened #17582 so we don't forget about it once this PR is merged.

pchote commented on OpenRA.Mods.Common/ColorValidator.cs:
```diff
@@ -94,7 +94,7 @@ public Color RandomPresetColor(MersenneTwister random, IEnumerable<Color> terrai
 			{
 				Color forbidden;
-				Action<string> ignoreError = _ => { };
+				HashSet<string> ignoreError = new HashSet<string>();
```
Two style issues here:
* Prefer `var` over explicit type names
* Variable name no longer matches usage

IMO the best fix for both issues would be to change `IsValid` to include null check before adding to the HashSet and give a default argument value of null. We can then remove the dummy set here completely.

pchote commented on OpenRA.Mods.Common/ColorValidator.cs:
```diff
@@ -107,7 +107,7 @@ public Color RandomValidColor(MersenneTwister random, IEnumerable<Color> terrain
 			Color color;
 			Color forbidden;
-			Action<string> ignoreError = _ => { };
+			HashSet<string> ignoreError = new HashSet<string>();
```
Likewise here.

pchote commented on OpenRA.Mods.Common/ColorValidator.cs:
```diff
@@ -122,8 +122,10 @@ public Color RandomValidColor(MersenneTwister random, IEnumerable<Color> terrain
 
 		public Color MakeValid(Color askedColor, MersenneTwister random, IEnumerable<Color> terrainColors, IEnumerable<Color> playerColors, Action<string> onError)
 		{
+			HashSet<string> errorMessages = new HashSet<string>();
```
`var errorMessages = new HashSet<string>();`

pchote commented:
LGTM otherwise, and it makes sense to get this in for the playtest so adding to the milestone.

Given the time constraints on the playtest release, I'll apply the fixups and merge this next weekend if not fixed sooner.

robpvn committed:
Run every color validation together

This ensures that color picks that have multiple issues will
have them all checked at the same time, including ensuring that
the fix for one issue doesn't cause another issue.

Handling of the onError action has been changed from being called
at once to collecting the potential errors in a HashSet to deduplicate
them and then calling onError after a valid color has been found.
(Otherwise you would in the worst case get 256 error messages logged!)

pchote reviewed (approved):
Pushed fixups, LGTM now :+1:

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/eea5648b93b6fec926c1d229c9420ffefaa32a43

robpvn commented:
Sorry about not responding promptly, I was laid up with the flu last week. Thanks for the merge!
